tests/admin/views/test_releases_v2.py

```diff
@@ -0,0 +1,178 @@
+from copy import deepcopy
+
+import pytest
+
+from auslib.global_state import dbo
+from auslib.util.data_structures import infinite_defaultdict
+
+
+def insert_release(release_data, product):
+    release_data = deepcopy(release_data)
+    if "platforms" in release_data:
+        for pname, pdata in release_data["platforms"].items():
+            if "locales" in pdata:
+                for lname, ldata in pdata["locales"].items():
+                    dbo.release_details.t.insert().execute(name=release_data["name"], path=f".platforms.{pname}.locales.{lname}", data=ldata, data_version=1)
+                del release_data["platforms"][pname]["locales"]
+    dbo.releases_json.t.insert().execute(
+        name=release_data["name"], product=product, data=release_data, data_version=1,
+    )
+
+
+@pytest.fixture(scope="function")
+def releases_db(db_schema, firefox_56_0_build1, firefox_60_0b3_build1, cdm_17):
+    dbo.setDb("sqlite:///:memory:")
+    db_schema.create_all(dbo.engine)
+    dbo.permissions.t.insert().execute(permission="admin", username="bob", data_version=1)
+    dbo.permissions.user_roles.t.insert().execute(username="bob", role="releng", data_version=1)
+    dbo.productRequiredSignoffs.t.insert().execute(product="Firefox", channel="release", role="releng", signoffs_required=1, data_version=1)
+    dbo.rules.t.insert().execute(
+        rule_id=1, priority=100, product="Firefox", channel="release", mapping="Firefox-56.0-build1", update_type="minor", data_version=1
+    )
+    dbo.rules.t.insert().execute(
+        rule_id=2, priority=100, product="Firefox", channel="beta", mapping="Firefox-60.0b3-build1", update_type="minor", data_version=1
+    )
+    insert_release(firefox_56_0_build1, "Firefox")
+    insert_release(firefox_60_0b3_build1, "Firefox")
+    insert_release(cdm_17, "CDM")
+
+
+@pytest.mark.usefixtures("releases_db", "mock_verified_userinfo")
+def test_overwrite_fails_when_signoff_required(api, firefox_56_0_build1):
+    firefox_56_0_build1["detailsUrl"] = "https://newurl"
+    firefox_56_0_build1["platforms"]["Darwin_x86_64-gcc3-u-i386-x86_64"]["locales"]["de"]["buildID"] = "9999999999999"
+
+    old_data_versions = infinite_defaultdict()
+    old_data_versions["."] = 1
+    for pname, pdata in firefox_56_0_build1["platforms"].items():
+        for lname in pdata.get("locales", {}):
+            old_data_versions["platforms"][pname]["locales"][lname] = 1
+
+    ret = api.put("/v2/releases/Firefox-56.0-build1", json={"blob": firefox_56_0_build1, "old_data_versions": old_data_versions})
+    assert ret.status_code == 400
+
+
+@pytest.mark.usefixtures("releases_db", "mock_verified_userinfo")
+def test_overwrite_succeeds(api, firefox_60_0b3_build1):
```

Comment:
Potential improvement: We may make 3 different test cases to ensure the whole thing works:

1. A unit test that verifies the api level delegates the right call to the mocked (or stubbed) out service layer.
2. A unit test that makes sure the service level delegates the right call to the database
3. An integration test that shows the API correctly stored the change by `get`ting the resource after the `put`.

This way, the current test doesn't have to deal with both the API and the database and will be lighter to setup/write. What do you think?

Reply:
Interesting that you mention this! The main reason I wrote the tests in a way where all of the entry points were the API was to avoid coupling the implementation and the tests together so tightly, as they have been in the past. This was inspired by https://blog.cleancoder.com/uncle-bob/2017/10/03/TestContravariance.html, the summary of which is:
> The structure of the tests must not reflect the structure of the production code, because that much coupling makes the system fragile and obstructs refactoring. Rather, the structure of the tests must be independently designed so as to minimize the coupling to the production code.

That doesn't mean more tests might not be useful, just wanted to add some background/explanation.